File: pybind/feature/sift.cpp
#include "pybind/feature/sift.h"

#include "src/feature/sift.h"

void pybind_feature_sift(py::module &m) {
    NDArrayConverter::init_numpy();

    py::module m_submodule = m.def_submodule("feature");
    py::class_<SiftExtractionOptions> sift_extraction_options(m_submodule, "SiftExtractionOptions");
    py::detail::bind_default_constructor<SiftExtractionOptions>(sift_extraction_options);
    py::detail::bind_copy_functions<SiftExtractionOptions>(sift_extraction_options);
    sift_extraction_options
        .def_readwrite("num_threads", &SiftExtractionOptions::num_threads)
        .def_readwrite("use_gpu", &SiftExtractionOptions::use_gpu)
        .def_readwrite("gpu_index", &SiftExtractionOptions::gpu_index)
        .def_readwrite("max_image_size", &SiftExtractionOptions::max_image_size)
        .def_readwrite("max_num_features", &SiftExtractionOptions::max_num_features)
        .def_readwrite("first_octave", &SiftExtractionOptions::first_octave)
        .def_readwrite("num_octaves", &SiftExtractionOptions::num_octaves)
        .def_readwrite("octave_resolution", &SiftExtractionOptions::octave_resolution)
        .def_readwrite("peak_threshold", &SiftExtractionOptions::peak_threshold)
        .def_readwrite("edge_threshold", &SiftExtractionOptions::edge_threshold)
        .def_readwrite("estimate_affine_shape", &SiftExtractionOptions::estimate_affine_shape)
        .def_readwrite("max_num_orientations", &SiftExtractionOptions::max_num_orientations)
        .def_readwrite("upright", &SiftExtractionOptions::upright)
        .def_readwrite("darkness_adaptivity", &SiftExtractionOptions::darkness_adaptivity)
        .def_readwrite("domain_size_pooling", &SiftExtractionOptions::domain_size_pooling)
        .def_readwrite("dsp_min_scale", &SiftExtractionOptions::dsp_min_scale)
        .def_readwrite("dsp_max_scale", &SiftExtractionOptions::dsp_max_scale)
        .def_readwrite("dsp_num_scales", &SiftExtractionOptions::dsp_num_scales)
        .def_readwrite("normalization", &SiftExtractionOptions::normalization);
    py::enum_<SiftExtractionOptions::Normalization>(m_submodule, "Normalization")
        .value("L1_ROOT", SiftExtractionOptions::Normalization::L1_ROOT)
        .value("L2", SiftExtractionOptions::Normalization::L2)
        .export_values();

    py::class_<SiftExtract> sift_extract(m_submodule, "SiftExtract");
    py::detail::bind_default_constructor<SiftExtract>(sift_extract);
    py::detail::bind_copy_functions<SiftExtract>(sift_extract);
    sift_extract
        .def(py::init<const SiftExtractionOptions&>())
        .def("Reset", &SiftExtract::Reset)
        .def("getKeyPoints", &SiftExtract::getKeyPoints)
        .def("getDescriptors", &SiftExtract::getDescriptors)
        .def("ExtractSiftFeatures", &SiftExtract::ExtractSiftFeatures);
}

File: python/feature/sift.py
import sys
sys.path.append("../../build/")
import pyVO
import numpy as np
import cv2

class SiftExtractor(object):
    def __init__(self, max_image_size=3200, max_num_features=8192, num_octaves=4,
        first_octave=-1, octave_resolution=3, edge_threshold=10.0, estimate_affine_shape=False,
        max_num_orientations=2, upright=False, darkness_adaptivity=False, domain_size_pooling=False,
        dsp_min_scale=1.0/6.0, dsp_max_scale=3.0, dsp_num_scales=10, use_gpu=False,
        gpu_index="-1", peak_threshold=0.02/3.0):
        super(SiftExtractor, self).__init__()

        self.sift_extract_options = pyVO.feature.SiftExtractionOptions()
        self.sift_extract_options.max_image_size = max_image_size
        self.sift_extract_options.max_num_features = max_num_features
        self.sift_extract_options.num_octaves = num_octaves
        self.sift_extract_options.first_octave = first_octave
        self.sift_extract_options.octave_resolution = octave_resolution
        self.sift_extract_options.edge_threshold = edge_threshold
        self.sift_extract_options.estimate_affine_shape = estimate_affine_shape
        self.sift_extract_options.max_num_orientations = max_num_orientations
        self.sift_extract_options.upright = upright
        self.sift_extract_options.darkness_adaptivity = darkness_adaptivity
        self.sift_extract_options.domain_size_pooling = domain_size_pooling
        self.sift_extract_options.dsp_min_scale = dsp_min_scale
        self.sift_extract_options.dsp_max_scale = dsp_max_scale
        self.sift_extract_options.dsp_num_scales = dsp_num_scales
        self.sift_extract_options.use_gpu = use_gpu
        self.sift_extract_options.gpu_index = gpu_index
        self.sift_extract_options.peak_threshold = peak_threshold

        self.sift_extract = pyVO.feature.SiftExtract(self.sift_extract_options)

    def extract(self, image):
        assert len(image.shape) == 3
        if image.shape[2] > 1:
            gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        else:
            gray = image
        self.sift_extract.ExtractSiftFeatures(gray)
        keypoints = self.sift_extract.getKeyPoints()
        kps = []
        for kp in keypoints:
            kps.append([kp.x, kp.y])
        kps = np.array(kps)
        descriptors = self.sift_extract.getDescriptors()
        descriptors = np.array(descriptors)
        return kps, descriptors

if __name__ == "__main__":
    sift_extractor = SiftExtractor()
    image = cv2.imread("data/test1.jpg", -1)
    kps, descriptors = sift_extractor.extract(image)

    for kp in kps:
        x, y = kp[0], kp[1]
        cv2.circle(image, (int(x+0.5), int(y+0.5)), 2, (0, 0, 255), 1)

    cv2.imwrite("test/sifttest.jpg", image)